RFC: Add descriptive names to doctests

casey opened the pull request:
I wasn't sure if this needed an RFC or not. It's a rather minor feature, however, it is a user-visible change, and since I've already written one, I thought I'd open a PR. Feel free to close it if it doesn't require an RFC.

I did some supporting work on this in https://github.com/rust-lang/rust/pull/78429, which made code block info string parsing only split on `,` and whitespace, so that an attribute of the form `name=foo` would be possible, but never followed up with the actual RFC.

This RFC proposes the ability to annotate documentation test code blocks with metadata of the form `name=NAME`. The given name will be used to identify the documentation test when it is run by the test runner, in addition to the current information of test binary, module, and line number.

For example, this documention test:

    ```name=arithmetic
    assert_eq!(2 + 2, 4);
    ```

Assuming that it is in crate `foo`, in file `src/lib.rs`, in module
`bar`, on line 37, will produce this output when run:

       Doc-tests foo

    running 1 test
    test src/lib.rs - bar::arithmetic (line 37) ... ok

    test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

[**Rendered**](https://github.com/casey/rfcs/blob/doctest-name/text/0000-doctest-name.md)

casey commented on text/0000-doctest-name.md:
```diff
@@ -0,0 +1,142 @@
+output by line number alone.
+
+By giving users the option to give names to doctests, identifying individual
+doctests in test runner output will be easier.
```
Ooooooo, nice, I didn't think of that.

Nemo157 commented on text/0000-doctest-name.md:
```diff
@@ -0,0 +1,142 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None.
```
Given the `"name="IDENTIFIER` syntax it's unambiguously the name `basic` in the `arithmetic` language (all 3 separators `[',', ' ', '\t']` cannot be part of an identifier).

`<IDENT>` being a valid rust identifier is only mentioned in the guide section though. It should probably be in the reference (and could link through to the definition too <https://doc.rust-lang.org/nightly/reference/identifiers.html>).

It may be worth mentioning that whitespace around the `=` is not allowed because of the existing separator support.

casey commented:
Updated and addressed comments. Thank you for the comments!

- Mentioned test filtering
- Remove stray note that `NAME` be a valid identifier
- Rename IDENT to NAME

A previous version of this RFC required that doctest names be valid rust identifiers. I think this might be a good idea, but someone wasn't convinced. I think it's probably a reasonable restriction, and means that you can use the identifiers as names for the generated test binaries/functions/etc, so it's something that I'd like feedback on. However, I wanted the test of the RFC to be consistent, so I removed references to that restriction for now.

joshtriplett commented on text/0000-doctest-name.md:
````diff
@@ -0,0 +1,143 @@
+```
+
+If you'd like to have the test runner print a more informative name, perhaps
+because there are many doctests in a single file, you can put `name=NAME`,
+where `NAME` is the name the test should have, in the doctest's info string.
````
```suggestion
because there are many doctests in a single file, you can put `name=IDENT`,
where `IDENT` is the name the test should have, in the doctest's info string.
```

And likewise elsewhere in the document: let's explicitly specify that the name must be an identifier. This also disambiguates cases like `name=ident,should_panic`.

casey replied:
Good call, done!

casey commented on text/0000-doctest-name.md:
```diff
@@ -0,0 +1,143 @@
+[prior-art]: #prior-art
+
+Unit and integration tests have names. This change would bring doctests closer
+to parity, by allowing features like test filtering to work with doctests.
```
`{#IDENT}` seems a little clunkier to me, although it is more compatible. I believe it isn't compatible with Rust's current markdown block attribute splitting code, which only splits on commas and space. Spaces, Rust does split on, do seem to be ignored by most markdown flavors, so `rust name=test` works.

I changed the RFC to use spaces instead of commas, since that seems more broadly ignored by markdown parsers.

casey commented:
Thanks for the feedback! I added the requirement that names be valid rust identifiers, and used spaces to separate attributes in the examples instead of commas.

casey commented:
Friendly ping!

casey commented:
cc @GuillaumeGomez

Are commas the preferred way (or only?) way to separate annotations on a code block? I used spaces originally, since they were accepted by the parser, but I'm not sure if that changed.

I think that's the only change here that needs to be made.

GuillaumeGomez commented:
Oh damn, completely forgot to stabilize this feature. Gonna send a PR for that.

Commas are the only character to separate annotations currently. To add your feature, I think you just need to check for `name` in [here](https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/markdown.rs#L1363-L1375). Once done and ready, please write a comment here so we review it one last time. 👍

casey commented:
@GuillaumeGomez Nice! Okay, updated the RFC to use commas in the example about separating attribute names. I think this is ready to go.

GuillaumeGomez commented:
Considering https://github.com/rust-lang/rust/pull/124577 is about to get stabilized, I'm not sure if `name` is good enough now (is `name` supposed to be an attribute or a doctest name? Can't know unless you know very well how it works).

Another thing to note is that you will be able to add attributes to generated codeblocks with this syntax: `custom,{name=something}`, meaning that in your case, `name` here would not name the doctest but instead be an attribute generated in the HTML. So better underline this case in the RFC.

In any case, I think renaming `name` into `doctest_name` would be better. What do you think?

GuillaumeGomez commented:
Good catch, you can see that key-values only only certain values [here](https://github.com/rust-lang/rust/pull/110800/files#diff-9797963bcc8ee935b3ce76a245b60bed2d5045deb1f170accb86c14c88b27908R1275).

casey commented:
I personally think `name` is better, just because it's short, and the people who use this feature will likely name every doctest, so they'll be writing out the attribute a bunch of times. Another option would be `test=NAME`, which is also short.

GuillaumeGomez commented:
One thing to note is that the eBNF doesn't currently allow to have `key=value` outside of curly braces since we followed pandoc syntax. I suppose it's ok since @notriddle approved. ;)

GuillaumeGomez commented:
Then `name` is a valid element attribute, so I'm not sure we should keep using `name` for this case as it would also modify the generated HTML documentation.

fmease commented:
This is a very nice feature!

@rfcbot concern braced-unbraced

As mentioned above, we need to decide if we'd like extend the code block attr grammar to support bare `name=IDENT` or if we would like to use/extend the existing[^1] pandoc syntax `{...,name=IDENT,...}`.

@rfcbot concern raw-idents

I guess it's implied by the use of the word *Rust identifier* but I'd like to raise awareness that this includes raw identifiers (prefix: `r#`) which is good. I think we should call this out explicitly mention this in the RFC text.

@rfcbot concern hard-errors

> Multiple `name=IDENT` words may not appear in a single code block's info
string.

I'd like to see this clarified. I'm pretty sure we don't want to emit a hard error for duplicate names but merely a warn-by-default `rustdoc` lint for backward compatibility.

@rfcbot reviewed

[^1] Currently feature-gated under `custom_code_classes_in_docs`

casey commented:
Thank you for the review!

> @rfcbot concern braced-unbraced
> 
> As mentioned above, we need to decide if we'd like extend the code block attr grammar to support bare `name=IDENT` or if we would like to use/extend the existing[^1] pandoc syntax `{...,name=IDENT,...}`.

I personally would prefer the unbraced `name=IDENT` syntax. Correct me if I'm wrong, but the pandoc braced syntax is only used for adding classes to the generated HTML, and it's also rather verbose.

> @rfcbot concern raw-idents
> 
> I guess it's implied by the use of the word _Rust identifier_ but I'd like to raise awareness that this includes raw identifiers (prefix: `r#`) which is good. I think we should call this out explicitly mention this in the RFC text.

I noted that raw identifiers are allowed.

> @rfcbot concern hard-errors
> 
> > Multiple `name=IDENT` words may not appear in a single code block's info
> > string.
> 
> I'd like to see this clarified. I'm pretty sure we don't want to emit a hard error for duplicate names but merely a warn-by-default `rustdoc` lint for backward compatibility.

In the same vein, should `name=NOT_AN_IDENT` warn-by-default as well, and not be used as a test name?

fmease commented:
> In the same vein, should `name=NOT_AN_IDENT` warn-by-default as well, and not be used as a test name?

Yes indeed, I forgot to mention that.

> https://github.com/rust-lang/rfcs/pull/3311/commits/7f90cc4781a421d5bc5a26b486c1909a593b8602

@rfcbot resolve raw-idents

fmease commented:
> should [IDENT in `name=IDENT`] not be used as a test name?

That's a separate point. And I honestly don't know if it's feasible or worth impl'ing this, namely checking that doctest names are unique within the value namespace of the current module (including other doctest names).

casey commented:
> That's a separate point. And I honestly don't know if it's feasible or worth impl'ing this, namely checking that doctest names are unique within the "`#[test]` namespace" of the current module (including other doctest names). Yeah, maybe we should actually check this and emit a warn-by-default lint on name clashes.

Just updated the RFC. If the idea is that `name=IDENT` doctests will potentially be instantiated as `#[test] fn test() { … }` items, then they should both be unique across doctests, but also unique across all items in the current namespace.

notriddle commented:
This feature was discussed in the previous monthly meeting, and we decided that it needed better motivation before adding it.

https://rust-lang.zulipchat.com/#narrow/channel/393423-t-rustdoc.2Fmeetings/topic/2024-12-09/near/487096843

To estimate how common this need might be, I used Crater to measure how many crates had items with more than five doctests to an item. The answer, as it turns out, is 0.6%.

https://github.com/rust-lang/rust/pull/134530

notriddle commented:
@rfcbot fcp cancel

@rfcbot fcp reject

https://rust-lang.zulipchat.com/#narrow/channel/393423-t-rustdoc.2Fmeetings/topic/2025-01-13/near/493447654

notriddle commented:
@rfcbot fcp close
